design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,262 @@
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+      * [References](#references)
+
+## Summary
+
+Code with the ability to perform the comparison on inspected and ready hosts and update the host with labels containing matched profiles.
+
+## Motivation
+
+We need to validate and compare hardware configuration against host hardware details-section.
+
+### Goals
+
+Purpose is to add a new CRD `HardwareClassificationController` to hold the  hardware details that reconciles those by looking for host CRs that matches with hardware classification profile and add label as "BMO-OK" for matched profile into baremetal host baremetalHost_crd.yaml.
+
+## Proposal
+
+We are creating proposal based on the discussion with metal3 community on issue #351:
+https://github.com/metal3-io/baremetal-operator/issues/351
+
+We compared introspection data of Ironic with metal3 schema and we found that default  hardware
+configuration needs to be added by introducing a new CRD `HardwareClassificationController`.
+
+We will write a CRD that checks inspected baremetal hosts against hardware configuration and add label as "BMO-OK" for matched profile into baremetal host baremetalHost_crd.yaml.
+  
+### Implementation Details/Notes/Constraints
+
+Link for Existing Metal3 Specs
+Please refer metal3 spec for bare-metal:
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+* Write a below schema for new CRD under folder deploy/crds for Kind HardwareClassificationController.
+
+    ```yaml
+    - properties:
+        expectedHardwareConfiguration:
+            default:
+            properties:
+            customFilter:
+                    type: String
+            namespace:
+                    type: String
+            profileName:
+                    type: String
+            cpu:
+                properties:
+                    count:
+                        type: Integer
+            type: Object
+            disk:
+                properties:
+                    rootDisk:
+                        sizeBytesGB:
+                            type: Integer
+                    type: Object
+                    dataDisk:
+                        sizeBytesGB:
+                            type: Integer
+                    numberOfDisks:
+                        type: Integer
+                    type: array
+            type: Object
+            nics:
+                properties:
+                    numberOfNICS:
+                        type: Integer
+            type: Object
+            ram:
+                properties:
+                    sizeBytesGB:
+                        type: Integer
+            type: Object
+            systemVendor:
+                properties:
+                    productName:
```

Comment:
Are the string fields going to require exact matches or will substrings match?

Reply:
We can discuss about it slack or on call.